### PR TITLE
Resample colorfill plot for faster interactive viewing.

### PR DESCRIPTION
**Pull request by quantumsteve**
Passing the entire workspace to matplotlib often leads to unacceptably slow performance. To work around this, we subclass AxesImage so that it first resizes the data to 4K, then resizes to the window size.

[scikit-image](http://scikit-image.org/docs/dev/api/skimage.transform.html#skimage.transform.resize) is used for the resampling. [scipy.misc.imresize](https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.imresize.html) fails with larger input arrays (integer overflow).

**Description of work.**

**To test:**

```python
import time
from mantid.simpleapi import *
import matplotlib.pyplot as plt
from mantid import plots
########## load/generate data

start = time.time()
w = LoadEventNexus(Filename='PG3_42283', OutputWorkspace='PG3_42283')
w = Rebin(InputWorkspace='PG3_42283', OutputWorkspace='PG3_42283', Params=1)
print('generation took {} (s)'.format(time.time() - start))

########## plot
start = time.time()
fig, ax = plt.subplots(subplot_kw={'projection': 'mantid'})
ax.imshow(w, cmap='viridis', interpolation='nearest')
ax.set_aspect('auto')
fig.set_tight_layout(True)
fig.show()
print('plot took ', time.time() - start, '(s)')
```

There is no associated issue.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -394,6 +394,7 @@ if ( ENABLE_CPACK )
                                            "python3-six,"
                                            "python3-matplotlib,"
                                            "python3-scipy,"
+					   "python3-skimage,"
```
Comment (peterfpeterson) [inline CMakeLists.txt]: match the existing spacing please

File: CMakeLists.txt [hunk 2]
```diff
@@ -413,6 +414,7 @@ if ( ENABLE_CPACK )
                                            "python-six,"
                                            "python-matplotlib,"
                                            "python-scipy,"
+					   "python-skimage"
```
Comment (peterfpeterson) [inline CMakeLists.txt]: match the existing spacing please

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Minor spacing change. Also need to revise the developer package versions.

Overall looks good and works well. Resizing the window happens near instantly. A later PR will have to worry about zoom/pan actions.

**Comment by peterfpeterson** [comment]
We'll need to install `python2-scikit-image` and `python3-scikit-image` for the builds to pass.

Also will need to deploy new developer packages.

**Comment by martyngigg** [comment]
@DTasev has created a commit for adding `scikit-image` to the Windows dependencies. You'll have to update [Bootstrap.cmake](https://github.com/mantidproject/mantid/blob/master/buildconfig/CMake/Bootstrap.cmake#L13) to `7c4c51e1a705d9dd05bf9b2d27dc102a41485c07` to pull it in.

File: CMakeLists.txt (hunk 1 above)
Comment (quantumsteve) [reply CMakeLists.txt]: Changed tabs to spaces

**Review by peterfpeterson — DISMISSED** [review]

File: MantidPlot/make_package.rb.in [hunk 3]
```diff
@@ -339,7 +339,7 @@ system_python_extras = "/System/Library/Frameworks/Python.framework/Versions/2.7
 pip_site_packages = "/Library/Python/2.7/site-packages"
 directories = ["sphinx","sphinx_bootstrap_theme","IPython","zmq","pygments","backports", "qtawesome", "qtpy",
                "certifi","tornado","markupsafe","matplotlib","mpl_toolkits", "jinja2","jsonschema","functools32",
-               "ptyprocess","CifFile","yaml","requests"]
+               "ptyprocess","CifFile","yaml","requests","skimage"]
```
Comment (martyngigg) [inline MantidPlot/make_package.rb.in]: Will this correctly pull in all of the dependencies?

Comment (quantumsteve) [reply MantidPlot/make_package.rb.in]: Looks like packaging will require more work...

Comment (quantumsteve) [reply MantidPlot/make_package.rb.in]: Added everything suggested by `homebrew-pypi-poet`.

**Comment by quantumsteve** [comment]
@martyngigg How should I fix the merge conflict in `Bootstrap.cmake`?

**Comment by martyngigg** [comment]
@quantumsteve Please replace the value with `80b72f1c5ad30675c03967ead71c8223fbe50d4f`.

We should merge mantidproject/thirdparty-msvc2015#41 when this pull request is merged.

**Comment by quantumsteve** [comment]
Had to add a check before enabling anti-aliasing. Old versions don't have and new versions give a warning if it isn't set.
http://scikit-image.org/docs/0.14.x/release_notes_and_installation.html

File: Framework/PythonInterface/mantid/plots/plotfunctions.py [hunk 4]
```diff
@@ -378,8 +378,18 @@ def pcolormesh(axes, workspace, *args, **kwargs):
     return axes.pcolormesh(x, y, z, *args, **kwargs)
 
 
+def _skimage_version():
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/plots/plotfunctions.py]: You can use `LooseVersionCheck` from `distutils` ([example](https://github.com/mantidproject/mantid/blob/76178489f6bafb8165682400f5c269f96516f4df/Framework/PythonInterface/test/python/plugins/algorithms/StringToPngTest.py#L25))

File: qt/applications/workbench/workbench/plotting/functions.py [hunk 5]
```diff
@@ -196,6 +196,23 @@ def pcolormesh_from_names(names, fig=None):
         return None
 
 
+def use_imshow(ws):
```
Comment (peterfpeterson) [inline qt/applications/workbench/workbench/plotting/functions.py]: Nice use of numpy

File: qt/applications/workbench/workbench/plotting/functions.py [hunk 6]
```diff
@@ -218,7 +235,10 @@ def pcolormesh(workspaces, fig=None):
         if subplot_idx < workspaces_len:
             ws = workspaces[subplot_idx]
             ax.set_title(ws.name())
-            pcm = ax.pcolormesh(ws, cmap=DEFAULT_CMAP)
+            if use_imshow(ws):
+                pcm = ax.imshow(ws, cmap=DEFAULT_CMAP, aspect='auto')
```
Comment (AndreiSavici) [inline qt/applications/workbench/workbench/plotting/functions.py]: Not a good idea. What if the user relies on getting a pcolormesh back?

Comment (quantumsteve) [reply qt/applications/workbench/workbench/plotting/functions.py]: Where would this be called outside of the GUI?

**Review by peterfpeterson — APPROVED** [review]
This works as advertised. One thing to not is the right-click menu still uses `pcolormesh` when the binning parameter is set to `10` because the last bin in the workspace is `6.801...` wide and doesn't match the other bins. I don't think the average user would care if the last bin had a distortion in it, but that is a matter for the follow-on work when this is made to work with non-constant binning.

**Pull request merged**